lib/Analysis/FIRRTLInstanceInfo.cpp

```diff
@@ -0,0 +1,156 @@
+//===- FIRRTLInstanceInfo.cpp - Instance info analysis ----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InstanceInfo analysis.  This is an analysis that
+// depends on the InstanceGraph analysis, but provides additional information
+// about FIRRTL operations.  This is useful if you find yourself needing to
+// selectively iterate over parts of the design.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Analysis/FIRRTLInstanceInfo.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-analysis-instanceinfo"
+
+using namespace circt;
+using namespace firrtl;
+using namespace detail;
+
+void InstanceInfo::LatticeValue::merge(LatticeValue that) {
+  if (kind > that.kind)
+    return;
+
+  if (kind < that.kind) {
+    kind = that.kind;
+    constant = that.constant;
+    return;
+  }
+
+  if (kind == Constant && constant != that.constant)
+    kind = Mixed;
+}
+
+void InstanceInfo::LatticeValue::merge(bool property) {
+  merge({/*kind=*/Constant, /*constant=*/property});
+}
+
+namespace {
+/// This struct is one frame of a worklist.  This is used to track what instance
+/// is being visited and with what information about its instantiation needs to
+/// be known.
+struct Frame {
+
+  /// The instance that will be visited.
+  InstanceGraphNode *node;
+
+  /// Indicates that the current hierarchy is under the DUT.
+  bool underDut = false;
+
+  /// Indicates that the current hierarchy is under a layer.
+  bool underLayer = false;
+};
+} // namespace
+
+InstanceInfo::InstanceInfo(Operation *op, mlir::AnalysisManager &am) {
```

Comment:
Is the `op` required here ?

Reply:
According to the docs [[1]], yes, it needs to have a constructor of `(Operation *)` or `(Operation *, AnalysisManager &)`. I tested this just now and the compiler will reject it without it.

Good eye, though. This is clearly unused in the body.

[1]: https://mlir.llvm.org/docs/PassManagement/#analysis-management